src/Microsoft.AspNetCore.Routing/EndpointConstraints/EndpointSelector.cs

```diff
@@ -0,0 +1,442 @@
+
+namespace Microsoft.AspNetCore.Routing.EndpointConstraints
+{
+    internal class EndpointSelector
+    {
+        private static readonly IReadOnlyList<Endpoint> EmptyEndpoints = Array.Empty<Endpoint>();
+
+        private readonly CompositeEndpointDataSource _dataSource;
+        private readonly EndpointConstraintCache _endpointConstraintCache;
+        private readonly ILogger _logger;
+
+        private Cache _cache;
+
+        public EndpointSelector(
+            CompositeEndpointDataSource dataSource,
+            EndpointConstraintCache endpointConstraintCache,
+            ILoggerFactory loggerFactory)
+        {
+            _dataSource = dataSource;
+            _logger = loggerFactory.CreateLogger<EndpointSelector>();
+            _endpointConstraintCache = endpointConstraintCache;
+        }
+
+        private Cache Current
+        {
+            get
+            {
+                var cache = Volatile.Read(ref _cache);
+
+                if (cache != null)
+                {
+                    return cache;
+                }
+
+                cache = new Cache(_dataSource.Endpoints);
+                Volatile.Write(ref _cache, cache);
+                return cache;
+            }
+        }
+
+        public IReadOnlyList<Endpoint> SelectCandidates(RouteContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var cache = Current;
+
+            // The Cache works based on a string[] of the route values in a pre-calculated order. This code extracts
+            // those values in the correct order.
+            var keys = cache.RouteKeys;
+            var values = new string[keys.Length];
+            for (var i = 0; i < keys.Length; i++)
+            {
+                context.RouteData.Values.TryGetValue(keys[i], out object value);
+
+                if (value != null)
+                {
+                    values[i] = value as string ?? Convert.ToString(value);
+                }
+            }
+
+            if (cache.OrdinalEntries.TryGetValue(values, out var matchingRouteValues) ||
+                cache.OrdinalIgnoreCaseEntries.TryGetValue(values, out matchingRouteValues))
+            {
+                Debug.Assert(matchingRouteValues != null);
+                return matchingRouteValues;
+            }
+
+            //_logger.NoEndpointsMatched(context.RouteData.Values);
+            return EmptyEndpoints;
+        }
+
+        public Endpoint SelectBestCandidate(HttpContext context, IReadOnlyList<Endpoint> candidates)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (candidates == null)
+            {
+                throw new ArgumentNullException(nameof(candidates));
+            }
+
+            var finalMatches = EvaluateEndpointConstraints(context, candidates);
```

Comment:
From what I understood in our email discussion, this would be the point where we run a bunch of 'policies' in an order, which essentially act like filters which hopefully would give a single endpoint after their run. The current code could probably be broken into individual policies. Looking at something like `IEndpointConstraintMetadata` on an endpoint would be an implementation detail of our default policy(s).
As a user I could associate any kind of metadata to an endpoint and when my policy gets invoked I look for my own metadata and make decisions.